lib/auditd/auditd_linux.go

```diff
@@ -0,0 +1,353 @@
+	disabled
+	enabled
+)
+
+// Client is auditd client.
+type Client struct {
+	conn NetlinkConnector
+
+	execName     string
+	hostname     string
+	systemUser   string
+	teleportUser string
+	address      string
+	ttyName      string
+
+	mtx     sync.Mutex
+	dial    func(family int, config *netlink.Config) (NetlinkConnector, error)
+	enabled featureStatus
+}
+
+// auditStatus represent auditd status.
+// Struct comes https://github.com/linux-audit/audit-userspace/blob/222dbaf5de27ab85e7aafcc7ea2cb68af2eab9b9/docs/audit_request_status.3#L19
+// and has been updated to include fields added to the kernel more recently.
+type auditStatus struct {
+	Mask                  uint32 /* Bit mask for valid entries */
+	Enabled               uint32 /* 1 = enabled, 0 = disabled */
+	Failure               uint32 /* Failure-to-log action */
+	Pid                   uint32 /* pid of auditd process */
+	RateLimit             uint32 /* messages rate limit (per second) */
+	BacklogLimit          uint32 /* waiting messages limit */
+	Lost                  uint32 /* messages lost */
+	Backlog               uint32 /* messages waiting in queue */
+	Version               uint32 /* audit api version number or feature bitmap */
+	BacklogWaitTime       uint32 /* message queue wait timeout */
+	BacklogWaitTimeActual uint32 /* message queue wait timeout */
+}
+
+// IsLoginUIDSet returns true if login UID is set, false otherwise.
+func IsLoginUIDSet() bool {
+	if !hasCapabilities() {
+		// Current process doesn't have system permissions to talk to auditd.
+		return false
+	}
+
+	client := NewClient(Message{})
+	defer func() {
+		if err := client.Close(); err != nil {
+			log.WithError(err).Warn("failed to create auditd client")
+		}
+	}()
+	// We don't need to acquire the internal client mutex as the connection is
+	// not shared.
+	if err := client.connectUnderMutex(); err != nil {
+		return false
+	}
+
+	enabled, err := client.isEnabledUnderMutex()
+	if err != nil || !enabled {
+		return false
+	}
+
+	loginuid, err := getSelfLoginUID()
+	if err != nil {
+		log.WithError(err).Debug("failed to read login UID")
+		return false
+	}
+
+	// if value is not set, logind PAM module will set it to the correct value
+	// after fork. 4294967295 (math.MaxUint32) is -1 converted to uint32.
+	return loginuid != math.MaxUint32
+}
+
+func getSelfLoginUID() (int64, error) {
+	data, err := os.ReadFile("/proc/self/loginuid")
+	if err != nil {
+		return 0, trace.ConvertSystemError(err)
+	}
+
+	loginuid, err := strconv.ParseInt(string(data), 10, 64)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	return loginuid, nil
+}
+
+// SendEvent sends a single auditd event. Each request create a new netlink connection.
+// This function does not send the event and returns no error if it runs with no root permissions.
+func SendEvent(event EventType, result ResultType, msg Message) error {
+	if !hasCapabilities() {
+		// Do nothing when not running as root.
+		return nil
+	}
```

Comment:
Should the `NewClient` API also include this check? Maybe `NewClient` can return a no-op client / an error, or `clt.SendMsg` can also do this check and return nil?